File: src/test/java/no/fintlabs/InstanceMapperTest.java
package no.fintlabs;

import no.fintlabs.gateway.instance.model.File;
import no.fintlabs.gateway.instance.model.instance.InstanceObject;
import no.fintlabs.model.acos.AcosDocument;
import no.fintlabs.model.acos.AcosInstance;
import no.fintlabs.model.acos.AcosInstanceElement;
import no.fintlabs.model.acos.AcosInstanceMetadata;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.http.MediaType;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class InstanceMapperTest {

    private AcosInstanceMapper acosInstanceMapper;
    Function<File, Mono<UUID>> persistFile;

    @BeforeEach
    void setup() {
        acosInstanceMapper = new AcosInstanceMapper();
        persistFile = mock(Function.class);
    }

    @Test
    void shouldMapToInstance() {
        AcosInstance acosInstance = AcosInstance
                .builder()
                .metadata(
                        AcosInstanceMetadata
                                .builder()
                                .formId("TEST0488")
                                .instanceId("100384")
                                .build()
                )
                .formPdfBase64("formPdfBase64Value")
                .elements(List.of(
                        AcosInstanceElement.builder().id("Fornavn").value("Ola").build(),
                        AcosInstanceElement.builder().id("Etternavn").value("Nordmann").build(),
                        AcosInstanceElement.builder().id("Fornavn2").value("Kari").build(),
                        AcosInstanceElement.builder().id("Etternavn2").value("Ødegård").build(),
                        AcosInstanceElement.builder().id("Ukedag").value(null).build(),
                        AcosInstanceElement.builder().id("Farge_pa_bil").value("Grønn").build()
                ))
                .documents(List.of(
                        AcosDocument.builder()
                                .name("vedleggImageNavn")
                                .type(MediaType.IMAGE_JPEG)
                                .encoding(null)
                                .base64("vedleggImageBase64Value")
                                .build(),
                        AcosDocument.builder()
                                .name("vedleggVideoNavn")
                                .type(MediaType.parseMediaType("video/mp4"))
                                .encoding(null)
                                .base64("vedleggVideoBase64Value")
                                .build()
                ))
                .build();

        File expectedSkjemaPdfFile = File
                .builder()
                .name("skjemaPdf")
                .sourceApplicationId(1L)
                .sourceApplicationInstanceId("100384")
                .type(MediaType.APPLICATION_PDF)
                .encoding("UTF-8")
                .base64Contents("formPdfBase64Value")
                .build();

        File expectedVedleggImageFile = File
                .builder()
                .name("vedleggImageNavn")
                .sourceApplicationId(1L)
                .sourceApplicationInstanceId("100384")
                .type(MediaType.IMAGE_JPEG)
                .encoding(null)
                .base64Contents("vedleggImageBase64Value")
                .build();

        File expectedVedleggVideoFile = File
                .builder()
                .name("vedleggVideoNavn")
                .sourceApplicationId(1L)
                .sourceApplicationInstanceId("100384")
                .type(MediaType.parseMediaType("video/mp4"))
                .encoding(null)
                .base64Contents("vedleggVideoBase64Value")
                .build();

        when(persistFile.apply(expectedSkjemaPdfFile)).thenReturn(Mono.just(UUID.fromString("391e9177-2790-469a-9f42-c8042731bc55")));
        when(persistFile.apply(expectedVedleggImageFile)).thenReturn(Mono.just(UUID.fromString("dab3ecc8-2901-46f0-9553-2fbc3e71ae9e")));
        when(persistFile.apply(expectedVedleggVideoFile)).thenReturn(Mono.just(UUID.fromString("5a15e2dd-29a7-41ac-a635-f4ab41d10d18")));

        InstanceObject instance = acosInstanceMapper.map(
                1L,
                acosInstance,
                persistFile
        ).block();

        InstanceObject expectedInstanceObject = InstanceObject
                .builder()
                .valuePerKey(Map.of(
                        "skjema.Fornavn", "Ola",
                        "skjema.Etternavn", "Nordmann",
                        "skjema.Fornavn2", "Kari",
                        "skjema.Etternavn2", "Ødegård",
                        "skjema.Ukedag", "",
                        "skjema.Farge_pa_bil", "Grønn",
                        "skjemaPdf", "391e9177-2790-469a-9f42-c8042731bc55"
                ))
                .objectCollectionPerKey(Map.of(
                        "vedlegg", List.of(
                                InstanceObject
                                        .builder()
                                        .valuePerKey(Map.of(
                                                "navn", "vedleggImageNavn",
                                                "type", "image/jpeg",
                                                "enkoding", "",
                                                "fil", "dab3ecc8-2901-46f0-9553-2fbc3e71ae9e"
                                        ))
                                        .build(),
                                InstanceObject
                                        .builder()
                                        .valuePerKey(Map.of(
                                                "navn", "vedleggVideoNavn",
                                                "type", "video/mp4",
                                                "enkoding", "",
                                                "fil", "5a15e2dd-29a7-41ac-a635-f4ab41d10d18"
                                        ))
                                        .build()
                        )
                ))
                .build();

        assertEquals(expectedInstanceObject, instance);
    }


}

File: src/test/kotlin/no/novari/flyt/acos/instance/gateway/InstanceMapperTest.kt
package no.novari.flyt.acos.instance.gateway

import no.novari.flyt.acos.instance.gateway.model.acos.AcosDocument
import no.novari.flyt.acos.instance.gateway.model.acos.AcosInstance
import no.novari.flyt.acos.instance.gateway.model.acos.AcosInstanceElement
import no.novari.flyt.acos.instance.gateway.model.acos.AcosInstanceMetadata
import no.novari.flyt.gateway.instance.model.File
import no.novari.flyt.gateway.instance.model.instance.InstanceObject
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.whenever
import org.springframework.http.MediaType
import java.util.UUID

@ExtendWith(MockitoExtension::class)
class InstanceMapperTest {
    private lateinit var acosInstanceMapper: AcosInstanceMapper

    @Mock
    private lateinit var persistFile: (File) -> UUID

    @BeforeEach
    fun setup() {
        acosInstanceMapper = AcosInstanceMapper()
    }

    @Test
    fun shouldMapToInstance() {
        val acosInstance =
            AcosInstance(
                metadata =
                    AcosInstanceMetadata(
                        formId = "TEST0488",
                        instanceId = "100384",
                    ),
                formPdfBase64 = "formPdfBase64Value",
                elements =
                    listOf(
                        AcosInstanceElement(id = "Fornavn", value = "Ola"),
                        AcosInstanceElement(id = "Etternavn", value = "Nordmann"),
                        AcosInstanceElement(id = "Fornavn2", value = "Kari"),
                        AcosInstanceElement(id = "Etternavn2", value = "Ødegård"),
                        AcosInstanceElement(id = "Ukedag", value = null),
                        AcosInstanceElement(id = "Farge_pa_bil", value = "Grønn"),
                    ),
                documents =
                    listOf(
                        AcosDocument(
                            name = "vedleggImageNavn",
                            type = MediaType.IMAGE_JPEG,
                            encoding = null,
                            base64 = "vedleggImageBase64Value",
                        ),
                        AcosDocument(
                            name = "vedleggVideoNavn",
                            type = MediaType.parseMediaType("video/mp4"),
                            encoding = null,
                            base64 = "vedleggVideoBase64Value",
                        ),
                    ),
            )

        val expectedSkjemaPdfFile =
            File(
                name = "skjemaPdf",
                sourceApplicationId = 1L,
                sourceApplicationInstanceId = "100384",
                type = MediaType.APPLICATION_PDF,
                encoding = "UTF-8",
                base64Contents = "formPdfBase64Value",
            )

        val expectedVedleggImageFile =
            File(
                name = "vedleggImageNavn",
                sourceApplicationId = 1L,
                sourceApplicationInstanceId = "100384",
                type = MediaType.IMAGE_JPEG,
                encoding = "",
                base64Contents = "vedleggImageBase64Value",
            )

        val expectedVedleggVideoFile =
            File(
                name = "vedleggVideoNavn",
                sourceApplicationId = 1L,
                sourceApplicationInstanceId = "100384",
                type = MediaType.parseMediaType("video/mp4"),
                encoding = "",
                base64Contents = "vedleggVideoBase64Value",
            )

        whenever(persistFile(expectedSkjemaPdfFile))
            .thenReturn(UUID.fromString("391e9177-2790-469a-9f42-c8042731bc55"))
        whenever(persistFile(expectedVedleggImageFile))
            .thenReturn(UUID.fromString("dab3ecc8-2901-46f0-9553-2fbc3e71ae9e"))
        whenever(persistFile(expectedVedleggVideoFile))
            .thenReturn(UUID.fromString("5a15e2dd-29a7-41ac-a635-f4ab41d10d18"))

        val instance =
            acosInstanceMapper.map(
                sourceApplicationId = 1L,
                incomingInstance = acosInstance,
                persistFile = persistFile,
            )

        val expectedInstanceObject =
            InstanceObject(
                valuePerKey =
                    mapOf(
                        "skjema.Fornavn" to "Ola",
                        "skjema.Etternavn" to "Nordmann",
                        "skjema.Fornavn2" to "Kari",
                        "skjema.Etternavn2" to "Ødegård",
                        "skjema.Ukedag" to "",
                        "skjema.Farge_pa_bil" to "Grønn",
                        "skjemaPdf" to "391e9177-2790-469a-9f42-c8042731bc55",
                    ),
                objectCollectionPerKey =
                    mutableMapOf(
                        "vedlegg" to
                            listOf(
                                InstanceObject(
                                    valuePerKey =
                                        mapOf(
                                            "navn" to "vedleggImageNavn",
                                            "type" to "image/jpeg",
                                            "enkoding" to "",
                                            "fil" to "dab3ecc8-2901-46f0-9553-2fbc3e71ae9e",
                                        ),
                                ),
                                InstanceObject(
                                    valuePerKey =
                                        mapOf(
                                            "navn" to "vedleggVideoNavn",
                                            "type" to "video/mp4",
                                            "enkoding" to "",
                                            "fil" to "5a15e2dd-29a7-41ac-a635-f4ab41d10d18",
                                        ),
                                ),
                            ),
                    ),
            )

        assertThat(instance).isEqualTo(expectedInstanceObject)
    }

    @Test
    fun shouldMapDuplicateElementIdsToLineSeparatedValues() {
        val acosInstance =
            AcosInstance(
                metadata =
                    AcosInstanceMetadata(
                        formId = "TEST0488",
                        instanceId = "100384",
                    ),
                formPdfBase64 = "formPdfBase64Value",
                elements =
                    listOf(
                        AcosInstanceElement(
                            id = "Rapport.Oppgi_navn_og1.0.Organisasjon2.Organisasjonsnavn1_1",
                            value = "FIKTIV SERVICE AS",
                        ),
                        AcosInstanceElement(
                            id = "Rapport.Oppgi_navn_og1.0.Organisasjon2.Organisasjonsnummer2_1",
                            value = "123456789",
                        ),
                        AcosInstanceElement(
                            id = "Rapport.Oppgi_navn_og1.0.Organisasjon2.Organisasjonsnavn1_1",
                            value = "FIKTIV1 AS",
                        ),
                        AcosInstanceElement(
                            id = "Rapport.Oppgi_navn_og1.0.Organisasjon2.Organisasjonsnummer2_1",
                            value = "987654321",
                        ),
                        AcosInstanceElement(
                            id = "Rapport.Oppgi_navn_og1.0.Organisasjon2.Organisasjonsnavn1_1",
                            value = "FIKTIV2 AS",
                        ),
                        AcosInstanceElement(
                            id = "Rapport.Oppgi_navn_og1.0.Organisasjon2.Organisasjonsnummer2_1",
                            value = "789456123",
                        ),
                    ),
            )

        val expectedSkjemaPdfFile =
            File(
                name = "skjemaPdf",
                sourceApplicationId = 1L,
                sourceApplicationInstanceId = "100384",
                type = MediaType.APPLICATION_PDF,
                encoding = "UTF-8",
                base64Contents = "formPdfBase64Value",
            )

        whenever(persistFile(expectedSkjemaPdfFile))
            .thenReturn(UUID.fromString("391e9177-2790-469a-9f42-c8042731bc55"))

        val instance =
            acosInstanceMapper.map(
                sourceApplicationId = 1L,
                incomingInstance = acosInstance,
                persistFile = persistFile,
            )

        assertThat(instance.valuePerKey)
            .containsEntry(
                "skjema.Rapport.Oppgi_navn_og1.0.Organisasjon2.Organisasjonsnavn1_1",
                "FIKTIV SERVICE AS\nFIKTIV1 AS\nFIKTIV2 AS",
            ).containsEntry(
                "skjema.Rapport.Oppgi_navn_og1.0.Organisasjon2.Organisasjonsnummer2_1",
                "123456789\n987654321\n789456123",
            )
    }
}
